Simpler API Blueprint Tutorial

kylef opened the pull request:
These changes make the new tutorial shorter, and simpler while requiring less prerequisites. The tutorial has no expectations of the user knowing Markdown.

Would appreciate any feedback on this one to ensure we get this right.

zdne commented on Tutorial.md:
```diff
@@ -1,7 +1,8 @@
+# API Blueprint Basic Tutorial
 
-Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set some expectations let's provide following functionality in our API: *List Gists, Retrieve a Single Gist, Create a Gist, Edit a Gist, Star a Gist, Unstar a Gist, Check if a Gist is starred and Delete a Gist.*
+Welcome to the API Blueprint Tutorial! In this tutorial you will learn about the basics to getting started with the API Blueprint language.
```
I am definitely not the native speaker here but shouldn't this be "about the basics to get started" ?

zdne replied:
Unity: I feel you are switching pronouns quite often is it "you learning" or "us learning" the API Blueprint? There is 18 occurrences of "our", 25 occurrences of "you", and one occurrence of "yours". 

I would say the tutorial should decide at what person it is "talking to" and stay with it. By that I mean it is either 

"In this tutorial you will learn" & "your blueprint"

Or

"In this tutorial we will learn" & "our blueprint"

Not the mix of the two. Personally I vote for the former as it is a little less formal

zdne commented on Tutorial.md:
```diff
@@ -1,7 +1,8 @@
-Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set some expectations let's provide following functionality in our API: *List Gists, Retrieve a Single Gist, Create a Gist, Edit a Gist, Star a Gist, Unstar a Gist, Check if a Gist is starred and Delete a Gist.*
+Welcome to the API Blueprint Tutorial! In this tutorial you will learn about the basics to getting started with the API Blueprint language.
+
+We’re going to build an API Blueprint for a service called polls, a simple web service that allows consumers to view polls and vote in them.
```
Again sorry for possibly being illiterate but should Polls be with capital P? "a service called polls" also I feel the dash is more appropriate "a service called Polls – a simple web service.."

zdne commented on Tutorial.md:
```diff
@@ -1,7 +1,8 @@
+
+We’re going to build an API Blueprint for a service called polls, a simple web service that allows consumers to view polls and vote in them.
 
 > **Note:** **Additional API Blueprint Resources**
```
This feels hefty at this point. Maybe we should move it to the tail of the tutorial and merge it with the content there?

zdne commented on Tutorial.md:
```diff
@@ -11,455 +12,240 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
-                        "gists": { "href": "/gists?{since}", "templated": true }
-                    }
-                }
+At the top of our blueprint, you will find a metadata section. In this case we have specified that "FORMAT" has the value of "1A". The format keyword is to specify the version, or format of API Blueprint we're using. In this case 1A.
```
Simplify: remove ", or format". Reminder unify pronouns "we're using" vs. "you are using"

zdne commented on Tutorial.md:
```diff
@@ -11,455 +12,240 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+At the top of our blueprint, you will find a metadata section. In this case we have specified that "FORMAT" has the value of "1A". The format keyword is to specify the version, or format of API Blueprint we're using. In this case 1A.
 
-What is going on here?
+The metadata section allows you to place information for use by [tooling around API Blueprint][Tooling Section]. For example, [Apiary.io](https://apiary.io/) uses a piece of metadata called `HOST`. When specified, it allows you to specify where an implementation of the API can be found.
```
While valid and correct I believe this adds unnecessary complexity at this point. I would remove this paragraph completely.

zdne commented on Tutorial.md:
```diff
@@ -11,455 +12,240 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+> **Note:** Should you need a clarification of some terms as used through this document refer to [API Blueprint Glossary of Terms][].
 
-#### Resource Description
+## Title & Description
```
(API) Name sounds better...

zdne commented on Tutorial.md:
```diff
@@ -11,455 +12,240 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+After the metadata we have a heading, headings start with a hash (#) followed by a title. In this case `# Polls`. You can use up to 6 hashes which allows you to have sub-headings. The first heading in our API Blueprint document serves as the API Name.
 
-    This resource does not have any attributes. Instead it offers the initial API affordances in the form of the HTTP Link header and haveL links.
+Then we have a description of our API. Where we have mentioned what this is. You can use further headings to break up the description sections.
```
Repeating "we have", probably concatenate those two sentences?

zdne commented on Tutorial.md:
```diff
@@ -11,455 +12,240 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+## Resource Groups
 
-#### Action
+Now it's time to start documenting our resources in our API. To start of with, we're going to can use the `Group` keyword inside a heading to specify a group of related resources. This allows us to group all of our resources related to questions in our API together.
```
repeating "related"

zdne replied:
"we're going to can" :arrow_right: "we're going to"

zdne commented on Tutorial.md:
```diff
@@ -11,455 +12,240 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+## Resource
 
-    + Response 200
+Inside of questions resource group, we have a resource in our API called "Questions collection". This resource allows you to view a list of questions. We've created a sub-heading inside our group called `Questions collection` which then specifies the URI for this resource inside our API which is placed inside square brackets at the end of the title.
```
Omit "in our API" it is a clutter

zdne replied:
Style unity sometimes you refer to names in quotes sometimes without quotes `called polls` vs. `called "Questions collection".

zdne replied:
Repeating "called"

zdne replied:
Remove "inside our API"

zdne commented on Tutorial.md:
```diff
@@ -11,455 +12,240 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
 
-    + Response 200 (application/hal+json)
-        + Headers
+In an API Blueprint, you need to specify the actions you can make on this resource in HTTP. For example, we can retrive the list of questions using the HTTP method `GET`.
```
> In an API Blueprint, you need to specify the actions you can make on this resource in HTTP

 This doesn't feel right. For few reasons:

1. you do need to, you can
2. only one action is enough
3. omit the mention of HTTP completely

The following sentence needs rephrasing. Throw "in an API blueprint" out. It servers no purpose. Reader already know this. It is a clutter.

What about something along:

"You (We?) can specify one or multiple actions afforded by a resource."...

zdne commented on Tutorial.md:
```diff
@@ -11,455 +12,240 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+In an API Blueprint, you need to specify the actions you can make on this resource in HTTP. For example, we can retrive the list of questions using the HTTP method `GET`.
 
-                Link: <http:/api.gistfox.com/>;rel="self",<http:/api.gistfox.com/gists>;rel="gists"
+We can create a sub-heading inside our resource with a title for our action, followed by the HTTP method to use inside the square brackets.
```
This is repeating what was written above.

zdne commented on Tutorial.md:
```diff
@@ -11,455 +12,240 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
-                        "gists": { "href": "/gists?{since}", "templated": true }
-                    }
-                }
+In API Blueprint, an action **should** always include at least one response representing the HTTP response message sent back in response to a HTTP request. The response should always include a [status code][] and possibly an additional [payload][].
```
remove "In API Blueprint", if this is not clear to the readers at this point then they are lost anyway :)

zdne replied:
unify the style of must, can, should etc (bold or not?)

zdne replied:
remove "additonal" the payload is not additional to anything

zdne commented on Tutorial.md:
```diff
@@ -11,455 +12,240 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+In API Blueprint, an action **should** always include at least one response representing the HTTP response message sent back in response to a HTTP request. The response should always include a [status code][] and possibly an additional [payload][].
 
-A response usually bears some payload sent to our client. Ideally it details the resource in question. In this case our response representation is of the [`application/hal+json`][] media-type. Also in addition to an HTTP response message-body this response defines an HTTP response [message-headers][] in its payload.
+For example, we can define the most common response "200" indicating the request has succeeded, along with an example result which is a JSON array of questions.
```
"the most common response "200" indicating the request has succeeded" is cool, true but a clutter, focus on teaching blueprint not HTTP.

zdne commented on Tutorial.md:
````diff
@@ -11,455 +12,240 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+        ]
+```
+
+> **Note:** API Blueprint is **indentation sensitive**. Assets have to be **indented by four spaces** relative to it's existing indentation level. In the case of list items or sections, which are already indented. It results in having you include additional indentation, this further makes it a total of **eight spaces** when using assets inside a list item.
````
The problem with this is that the word "asset" was never used before so a reader will be wondering "what the heck an asset is". Best to refer to it as payload or payload's body...

zdne replied:
"It results in having you include additional indentation" 

Doesn't make sense to me

zdne replied:
Probably best to remove the sentence about 8 spaces completely

zdne commented on Tutorial.md:
```diff
@@ -11,455 +12,240 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+> **Note:** API Blueprint is **indentation sensitive**. Assets have to be **indented by four spaces** relative to it's existing indentation level. In the case of list items or sections, which are already indented. It results in having you include additional indentation, this further makes it a total of **eight spaces** when using assets inside a list item.
 
-    + Response 200
+In our API, the polls resource has a second action allowing you to create a new question. Inside our action, we've added a description which describes the structure you would send to the server to create a new question.
```
"Inside our action" :arrow_right: "Inside this action"

zdne replied:
"description which describes" is truism, throw "which describes" out

zdne replied:
"the structure" :arrow_right: "an example of the structure" ?

zdne commented on Tutorial.md:
````diff
@@ -11,455 +12,240 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+```
 
-#### Request and its Payload
+We can show an example request, in this case we're showing an example JSON request you can make to the API.
````
the verb "show" does not feel right here

zdne commented on Tutorial.md:
```diff
@@ -11,455 +12,240 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
-    + Response 204
-
-In the case of the "Delete a Gist" action our response bears status code "204" indicating the server has successfully fulfilled the request and that there is no additional content to be send in the response payload body.
+This example request will result in a 201 response from the server specifying we've created a new resource. The API returns a `Location` header specifying the URI for the created resource along with a JSON body.
```
Reading this one loud does not feel right. Probably the "specifying" should be replaced. Or, better, do not teach HTTP here and skip explaining what 201 means.

zdne commented on Tutorial.md:
```diff
@@ -11,455 +12,240 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
-        + since (optional, string) ... Timestamp in ISO 8601 format: `YYYY-MM-DDTHH:MM:SSZ` Only Gists updated at or after this time are returned.
-
-    + Response 200
+The next resource in our group allows you to view the detail of a question directly with an ID of a question.
```
This is what bothers me. The resource is not allowing us something. Actions – affordances – are. List action allows us the view the detail. Not the resource itself. A nuance I know but feels important to me. The resource represents a data, an information. So a question is resource. "The next resource is Question ..."

zdne commented on Tutorial.md:
```diff
@@ -11,455 +12,240 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
-    The Star resource has the following attribute:
-
-    - starred
+The URI for “Question” resource uses a variable component, expressed by [URI Template][]. In this case, we have an id variable as `{question_id}`. Representing this resource is for `/questions/{question_id}` where `question_id` can be substituted for a real `id`.
```
> Representing this resource is for `/questions/{question_id}` where `question_id` can be substituted for a real `id`.

This reads really hard. I am not sure what you actually want to say here.

kylef replied:
Doesn't even make sense to the Kyle of Thursday.

![Double Facepalm](http://i.stack.imgur.com/jiFfM.jpg)

zdne commented on Tutorial.md:
```diff
@@ -11,455 +12,240 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
 
-    ### Check if a Gist is Starred [GET]
-    + Response 200
+You can describe a response without a body, for example a 204 response when we delete our question:
```
"Specifying a response without a body is simple as:"

zdne commented on Tutorial.md:
````diff
@@ -11,455 +12,240 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
++ Response 204
+```
 
 ## Complete Blueprint
````
I know, as with many things I am commenting here on, this goes on my own head, but: 

- API Blueprint: the name of the language
- API blueprint: a file – document – written in using the API Blueprint language 

"## Complete Blueprint" :arrow_right: "## Complete blueprint"

smizell replied:
@zdne Curious why the titles are not titlecase?

kylef replied:
I've left this as "Blueprint" with a capital B because that would be titlecase.

smizell replied:
:+1:

zdne commented on Tutorial.md:
```diff
@@ -11,455 +12,240 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
 
-    + Response 200 (application/hal+json)
-        + Headers
+With Poll Blueprint completed it is time to put it at work. For the start you can browse this API [rendered by GitHub][] or enjoy it [rendered by Apiary][].
```
I would elaborate here on additional resources. Also the name "API Blueprint Basic Tutorial" implies there is an advanced one...

zdne commented:
@kylef done with the review. Sorry for being maybe overly throughout but I feel this is, at the moment, the single most important learning material for API Blueprint (reserved for blueprint examples). Also I have probably spent too much time reading – [On Writing Well](http://www.amazon.com/Writing-Well-30th-Anniversary-Edition/dp/0060891548) (which I strongly recommend).

Simplify the most of the things. Throw away the clutter. Do not switch pronouns. Also I would throw any attempts to teach HTTP out of this document.

zdne commented:
Also, as this is a rewrite instead of a fresh new tutorial, I feel this is strongly influenced by my previous crappy writing. Sorry about that :)

kylef commented:
@zdne I think I've addressed most of your feedback. Thanks for going over it.

@smizell Would love you to take a quick read though if you have time too.

zdne commented on Tutorial.md:
```diff
@@ -1,7 +1,8 @@
-Welcome to the API Blueprint Tutorial! In this tutorial you will learn all about all the key features of the API Blueprint Language on the example of an API for our imaginary industry-changing Gist service.
 
-Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set some expectations let's provide following functionality in our API: *List Gists, Retrieve a Single Gist, Create a Gist, Edit a Gist, Star a Gist, Unstar a Gist, Check if a Gist is starred and Delete a Gist.*
+Welcome to an API Blueprint Tutorial! In this tutorial you will learn about the basics of the API Blueprint language.
```
"In this tutorial" repeat twice

zdne commented on Tutorial.md:
```diff
@@ -11,477 +12,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+### Actions
 
-Here we have just defined an action named "Retrieve Entry Point" utilizing the `GET` [HTTP Request Method][]. As with a resource and API name we can add any arbitrary description here but since the action name is pretty self explanatory we will skip it for now.
+API Blueprint allows you to specify the actions you can make on each resource. A sub-heading inside of the resource with a title, followed by the HTTP method to use inside the square brackets.
```
>  A sub-heading inside of the resource with a title, followed by the HTTP method to use inside the square brackets.

Reading this out loud: this seems very complex and I am actually not quite sure what it means as there is no active verb..

smizell replied:
The phrase quoted by @zdne is not a complete sentence. I also prefer "within" to inside as well, but that's not a hard rule. Additionally, I think it would be good to point out an example of what you are describing is below this paragraph.

zdne commented on Tutorial.md:
```diff
@@ -11,477 +12,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
-                        "gists": { "href": "/gists?{since}", "templated": true }
-                    }
-                }
+Action's should always include at least one response representing the HTTP response message sent back in response to a HTTP request. The response should always include a [status code][] and possibly a body.
```
Action's – 's shouldn't be here  (?)

zdne replied:
Remove the second word "response" (in "HTTP response")

zdne replied:
"and possibly a body." clutter (?)

smizell replied:
Does "should always" convey a MUST or a SHOULD? I would sync up this language with the spec. If MUST, use "Actions must include..." If SHOULD, "Actions should include..."

There are too many responses here as well (as noted by @zdne). I would even reword "in response to" to not have "response" in it.

Lastly, you have another "should always" in the last sentence. Sync this up with spec language, too.

kylef replied:
Gone for:

> An action should include at least one response from the server which must include a status could and may contain a body.

smizell replied:
I think there may be a typo here. 

> which must include a status could and may

zdne commented on Tutorial.md:
```diff
@@ -11,477 +12,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
 
-> **Note:** API Blueprint is **indentation sensitive**. See the [Note on Indentation][] for details.
->
 > **Note:** Specifying the media type in parenthesis after the response status codes generates implicit `Content-Type` HTTP header. Therefore you do not have to explicitly specify the `Content-Type` header.
```
Not sure about rendering this in Apiary – but reading the markdown version, this feels like it should be AFTER the example

smizell replied:
Agree. You should either say, "the following example..." or put it below the example.

zdne commented on Tutorial.md:
````diff
@@ -11,477 +12,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+```
 
-    + Response 200
+The polls resource has a second action, allowing you to create a new question. Inside this action, there is a description for the structure you would send to the server to create a new question.
````
Remove second "to create a new question"

smizell replied:
Have you introduced the idea of a "polls resource" yet? 

Also, the first sentence could be simplified:

> The polls resource has a second action which allows you to create a new question.

Lastly, "there is a description of the structure" has a **lot** in it, as it encapsulates MSON. Are you sure you want to pass over this so simply?

zdne commented on Tutorial.md:
```diff
@@ -11,477 +12,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
-    The Star resource has the following attribute:
-
-    - starred
+The URI for the “Question” resource uses a variable component, expressed by [URI Template][]. In this case, we have an id variable called `{question_id}. Where `question_id` is substituted for a real `id`.
```
missing closing backtick

zdne replied:
I would consider using "matched with" instead of "substituted for"

smizell replied:
Maybe something like:

> In this case, there is an ID variable called `question_id`, represented in the URI template as `{question_id}`. 

Do you need the last sentence? What is a "real ID?" 

Lastly, might want to sync on lowercase vs. uppercase `ID`.

zdne commented on Tutorial.md:
````diff
@@ -11,477 +12,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+```
 
-    + Model (application/hal+json)
+The "`question_id` variable of the URI template is a parameter for every action on this resource. Here defined of an arbitrary type `number`, followed by a description for the parameter.
````
extra `"`

smizell replied:
"It is defined here using an..."

zdne commented on Tutorial.md:
```diff
@@ -11,477 +12,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
-    ### List All Gists [GET]
-    + Parameters
-        + since (optional, string) ... Timestamp in ISO 8601 format: `YYYY-MM-DDTHH:MM:SSZ` Only Gists updated at or after this time are returned.
+The next resource is “Question”, it allows you to view a single question directly with an ID of a “Question”.
```
"an ID of a “Question”" :arrow_right "its ID" or "using its ID", but frankly I would not mention it (the ID) here 

Something like:

"The next resource is “Question”, which it allows you to view a single question."

Or better, if you do not need to make a claim about this resource actions: 

"The next resource is “Question”, which represents a single question."

smizell replied:
I agree with above, but also note if you leave in current form it is a run-on sentence. You need to end the first sentence after:

> The next resource is "Question."

zdne commented on Tutorial.md:
```diff
@@ -11,477 +12,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+### Actions
 
-                Link: <http:/api.gistfox.com/gists/42/star>;rel="self"
+Similar to before, this resource has an action to view the question detail.
```
At this point – reading really slowly and trying to get the grasp on what is API Blueprint – I do not recall what you are referring to in "similar to before" (remove?)

zdne replied:
This resource offers an action to retrieve it?

smizell replied:
View may work since it is titled below as "View."

kylef replied:
Good point, back when I wrote this sentence I didn't have a `URI Template`/`URI Parameter` section.

zdne commented:
2nd review done, yet I would like @smizell opinion as well

smizell commented on Tutorial.md:
```diff
@@ -1,7 +1,8 @@
-Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set some expectations let's provide following functionality in our API: *List Gists, Retrieve a Single Gist, Create a Gist, Edit a Gist, Star a Gist, Unstar a Gist, Check if a Gist is starred and Delete a Gist.*
+Welcome to an API Blueprint Tutorial! In this tutorial you will learn about the basics of the API Blueprint language.
+
+In this tutorial, we’re going to build an API Blueprint for a service called Polls – a simple web service that allows consumers to view polls and vote in them.
```
Do you want to refer to this as a Polls API rather than a service called Polls? Do you want to call this a "web service?"

I would replace "that" with something as well. "A simple web service for viewing polls and casting votes." Not saying you should use this example, but "that" is usually too general of a word for me. This may be mostly preference, too.

smizell commented on Tutorial.md:
```diff
@@ -11,477 +12,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
 
-    # Gist Fox API Root [/]
-    Gist Fox API entry point.
+The start of the Blueprint is going to look as follows:
```
When you say "the start," do you mean the first iteration of the blueprint for this tutorial, or the beginning of the document?

> The start of the Blueprint looks as follows:

I would remove "is going to."

Also, are we capitalizing "Blueprint" when it is without the "API" on the front?

smizell commented on Tutorial.md:
```diff
@@ -11,477 +12,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
 
-    + Response 200 (application/hal+json)
-        + Headers
+Polls is a simple web service that allows consumers to view polls and vote in them.
```
Ditto here https://github.com/apiaryio/api-blueprint/pull/154/files#r25521481

smizell commented on Tutorial.md:
```diff
@@ -11,477 +12,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
-                        "gists": { "href": "/gists?{since}", "templated": true }
-                    }
-                }
+At the very top of this Blueprint, you will find a metadata section. In this case we have specified that `FORMAT` has the value of `1A`. The format keyword is used to specify the version of the API Blueprint. In this case 1A.
```
I would sync up your language with my [comment here](https://github.com/apiaryio/api-blueprint/pull/154/files#r25521731) about "starting" or "beginning." It seems like the language should be beginning, middle, end rather than top, middle, bottom. You usually refer to writing as the former, maybe?

smizell commented on Tutorial.md:
```diff
@@ -11,477 +12,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+## API Name & Description
 
-#### Resource
+After the metadata we have a heading, headings start with a hash (#) followed by a title. In this case `# Polls`. You can use up to 6 hashes which allows you to have sub-headings. The first heading in the Blueprint document serves as the API Name.
```
First, break apart this sentence and add comma.

> After the metadata, we have a heading.

Also, would it be clearer to refer to `# Polls` first and then describe a title? I'm not sure "you can have up to six hashes" conveys the idea of nesting and sub-headings. 

Here is just an example to explain my last two issues:

> The first heading in the Blueprint document serves as the name of your API, which in this case is "Polls." Headings  start with at least one hash (#) followed by a title, and can be found throughout the Blueprint document. The API Name here uses one hash to distinguish it as the first level. Sub-headings may be used by adding more hashes to the heading, and can provide up to six levels of headings.

Also, if you keep "...6 hashes which..." please spell out six. You usually spell out number 10 or less.

Lastly, here you refer to this as an API Name, but above you've called this a web service. It would be good to sync up these terms.

smizell commented on Tutorial.md:
```diff
@@ -11,477 +12,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+After the metadata we have a heading, headings start with a hash (#) followed by a title. In this case `# Polls`. You can use up to 6 hashes which allows you to have sub-headings. The first heading in the Blueprint document serves as the API Name.
 
-    # Gist Fox API Root [/]
+Afterwards, there is a description of the Polls API, this mentions what the API is used for. You can use further headings to break up the description sections.
```
1. Afterwards is now meant to be a time-based word, so maybe a different word to indicate the location to which you are pointing.
2. Here you refer to this as Polls API, which you called just Polls above
3. After Polls API, end the sentence and start a new one
4. It is accepted now to end sentences with prepositions, but it seems informal to do so. "This mentions for what the API is used," is more formal of way to say it. Personally, I'd change this so it sounded better and didn't end with "for."
5. The last sentence feels like it could be more descriptive. Something like, "Headings may be used throughout the description."

I would take as a rule to not use "can" but rather [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt) language (even if not used explicitly). This tutorial should convey the constraints imposed by the API Blueprint spec, IMHO.

smizell replied:
I'm trying to stop using "can" myself in specs, and it's a hard habit to break :)

smizell commented on Tutorial.md:
```diff
@@ -11,477 +12,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+## Resource Groups
 
-> **NOTE:** Resource name is an [API Blueprint Identifier][] and as such it can be composed only from a combination of alphanumerical characters, underscores, dashes and a spaces.
+Now it's time to start documenting the resources inside of the API. To start of with, the `Group` keyword is used inside a heading to specify a group of related resources.
```
Inside is used twice. I would argue to use a different phrase than "inside of the API." Something like, "Now it's time to start documenting the API resources." 

Not sure you need "To start with."

smizell commented on Tutorial.md:
```diff
@@ -11,477 +12,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+## Resource
 
-As with the API Name the resource name can be followed by an arbitrary Markdown formatted description.
+Inside of the questions resource group, we have a resource called "Questions collection". This resource allows you to view a list of questions. The heading specifies the URI used to access the resource inside of square brackets at the end of the heading.
```
I would use "within" here instead of "inside." 

Two thoughts here:

1. Do you want to capitalize "collections," making it "Questions Collection?"
1. I'm not sure here, but is this a "Question Collection" or a "Questions Collection?" Which sounds better to you?

smizell commented on Tutorial.md:
```diff
@@ -11,477 +12,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
 
-    ### Delete a Gist [DELETE]
-    + Response 204
+You can create your own question using this action. It takes a JSON dictionary containing a question and a collection of answers in the form of choices.
```
First, "You **may** create..."

Also, it is a "JSON dictionary?" Isn't it a "JSON document?"

kylef replied:
After scanning the [JSON Specification](http://www.ietf.org/rfc/rfc4627.txt) they never use `document` nor `dictionary` to refer to any JSON element, but instead `object`. Will use `object` instead.

>   An object is an unordered collection of zero or more name/value pairs, where a name is a string and a value is a string, number, boolean, null, object, or array.

smizell replied:
Ha, I actually meant object. I am not sure why I typed document, but glad you figured it out.

smizell commented on Tutorial.md:
```diff
@@ -11,477 +12,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
-
-
-    + Model (application/hal+json)
+This example returns a 201 status code, along with HTTP headers and a JSON Body.
```
Above you had `200` as inline code but here you have 201 as just text. I'd vote include code again. Additionally, no need to capitalize "Body."

smizell commented on Tutorial.md:
````diff
@@ -11,477 +12,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
 
-        + Body
+```markdown
+### View a questions detail [GET]
````
Title-case the title?

smizell commented on Tutorial.md:
```diff
@@ -11,477 +12,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
 
-    ### Unstar a Gist [DELETE]
-    + Response 204
+#### Response without a body
```
Titlecase?

smizell commented on Tutorial.md:
```diff
@@ -11,477 +12,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
 
-    + Response 200 (application/hal+json)
-        + Headers
+Visit the [Tooling Section][] of [apiblueprint.org][] to find tools that you can use with API Blueprints.
```
"...tools to use with..."

smizell commented:
@kylef @zdne Review done. Please let me know when you make the additional changes and I'll read through it again.

kylef commented:
@smizell I've made a bunch of changes addressing your feedback.

smizell commented on Tutorial.md:
```diff
@@ -1,7 +1,6 @@
-Welcome to the API Blueprint Tutorial! In this tutorial you will learn all about all the key features of the API Blueprint Language on the example of an API for our imaginary industry-changing Gist service.
 
-Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set some expectations let's provide following functionality in our API: *List Gists, Retrieve a Single Gist, Create a Gist, Edit a Gist, Star a Gist, Unstar a Gist, Check if a Gist is starred and Delete a Gist.*
+Welcome to an API Blueprint Tutorial! This tutorial will take you though the basics of the API Blueprint language. We’re going to build an API Blueprint for a service called Polls – a simple API allowing consumers to view polls, and vote in them.
```
Remove comma after "polls."

> ...to view polls and vote in them.

smizell commented on Tutorial.md:
```diff
@@ -11,477 +10,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+## Metadata
 
-Depending on the tools you will be using with your blueprint you might want to start your blueprint file with a metadata explicitly stating the API Blueprint version you want to use like in this case the version 1A.
+The blueprint starts wih a metadata section. In this case we have specified that `FORMAT` has the value of `1A`. The format keyword is used to specify the version of the API Blueprint. In this case 1A.
```
Remove "that."

> ...we have specified `FORMAT` has...

Also, specify is used twice. You may make it something like:

> The format keyword denotes the version of the API Blueprint, which in this case is `1A`.

Be sure to make `1A` inline code like your first example of it.

smizell commented on Tutorial.md:
```diff
@@ -11,477 +10,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
 
-    + Response 200 (application/hal+json)
-        + Headers
+The first heading in the blueprint serves as the name of your API, which in this case is "Polls". Headings start with one or more `#` symbols followed by a title. The API Name here uses one hash to distinguish it as the first level. The number of `#` you use will determine the size of the heading.
```
Move period within quotation marks.

> ...which in this case is "Polls."

smizell replied:
Is it the "size of the heading" or is it the level of the heading (or something similar)?

smizell commented on Tutorial.md:
```diff
@@ -11,477 +10,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+The first heading in the blueprint serves as the name of your API, which in this case is "Polls". Headings start with one or more `#` symbols followed by a title. The API Name here uses one hash to distinguish it as the first level. The number of `#` you use will determine the size of the heading.
 
-                Link: <http:/api.gistfox.com/>;rel="self",<http:/api.gistfox.com/gists>;rel="gists"
+Following the heading is a description of the API, this mentions what the service is used for. You may use further headings to break up the description section.
```
End sentence after API.

> ...a description of the API. This mentions...

Also sentence ends in a preposition, "for."

smizell commented on Tutorial.md:
```diff
@@ -11,477 +10,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+## Resource
 
-#### Action
+Within the questions resource group, we have a resource called "Question Collection". This resource allows you to view a list of questions. The heading specifies the URI used to access the resource inside of square brackets at the end of the heading.
```
Move period within quotation>

> ...called "Question Collection."

smizell commented on Tutorial.md:
```diff
@@ -11,477 +10,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+## Metadata
 
-Depending on the tools you will be using with your blueprint you might want to start your blueprint file with a metadata explicitly stating the API Blueprint version you want to use like in this case the version 1A.
+The blueprint starts wih a metadata section. In this case we have specified that `FORMAT` has the value of `1A`. The format keyword specifies the version of the API Blueprint. In this case 1A.
```
Specify is used twice. You may make it something like:

> The format keyword denotes the version of the API Blueprint, which in this case is `1A`.

Be sure to make `1A` inline code like your first example of it.

smizell commented on Tutorial.md:
```diff
@@ -11,477 +10,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+## API Name & Description
 
-#### API Name
+The first heading in the blueprint serves as the name of your API, which in this case is "Polls". Headings start with one or more `#` symbols followed by a title. The API Name here uses one hash to distinguish it as the first level. The number of `#` you use will determine the size of the heading.
```
Move period within quotation marks.

> ...which in this case is "Polls."

Also, is it the "size of the heading" or is it the level of the heading (or something similar)?

smizell commented on Tutorial.md:
```diff
@@ -11,477 +10,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+The first heading in the blueprint serves as the name of your API, which in this case is "Polls". Headings start with one or more `#` symbols followed by a title. The API Name here uses one hash to distinguish it as the first level. The number of `#` you use will determine the size of the heading.
 
-    # Gist Fox API
+Following the heading is a description of the API, this mentions what the service is used for. You may use further headings to break up the description section.
```
End sentence after API.

> ...a description of the API. This mentions...

Also sentence ends in a preposition, "for."

smizell commented on Tutorial.md:
```diff
@@ -11,477 +10,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+## Resource
 
-#### Resource
+Within the questions resource group, we have a resource called "Question Collection". This resource allows you to view a list of questions. The heading specifies the URI used to access the resource inside of square brackets at the end of the heading.
```
Move period within quotation>

> ...called "Question Collection."

smizell commented on Tutorial.md:
```diff
@@ -11,477 +10,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+### Actions
 
-> **NOTE:** Resource name is an [API Blueprint Identifier][] and as such it can be composed only from a combination of alphanumerical characters, underscores, dashes and a spaces.
+API Blueprint allows you to specify the actions you can make on each resource. A sub-heading inside of the resource with a title, followed by the HTTP method to use inside the square brackets.
```
The last sentence is a sentence fragment. It needs a verb/predicate.

smizell commented on Tutorial.md:
```diff
@@ -11,477 +10,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
 
-    # Gist Fox API Root [/]
-    Gist Fox API entry point.
+An action should include at least one response from the server which must include a status could and may contain a body.
```
Typo?

> An action should include at least one response from the server which must include a **status could** and may contain a body.

smizell commented on Tutorial.md:
```diff
@@ -11,477 +10,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
 
-## Complex Resource
-With the entry point of our API defined we can move forward. Since our API revolves around a Gist let's define its representation and an action to retrieve it:
+> **Note:** Specifying the media type after the response status code generates `Content-Type` HTTP header. You do not have to explicitly specify the `Content-Type` header.
```
> ...status code generates **a** `Content-Type`...

smizell commented on Tutorial.md:
```diff
@@ -11,477 +10,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
-Here we have just defined a **Group** of our Gist-related resources. We have also defined a **Gist resource**, its **model representation** and **an action** that retrieves a single Gist.
-
-Let's break it down to what's new:
+The next resource is “Question”, which represents a single question.
```
Move comma inside quotation.

> ...is "Question," which...

smizell commented on Tutorial.md:
```diff
@@ -11,477 +10,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
 
-<a name="resource_model"></a>
-#### Resource Model
+The "`question_id` variable of the URI template is a parameter for every action on this resource. It's defined here using an arbitrary type `number`, followed by a description for the parameter.
```
Extra quotation mark before `question_id`.

smizell commented on Tutorial.md:
```diff
@@ -11,477 +10,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
 
-<a name="query_parameters"></a>
-#### Query Parameters
+This resource has a delete action, the server will return a 204 response without a body.
```
End sentence

> ...has a delete action. The server will...

smizell commented on Tutorial.md:
```diff
@@ -11,477 +10,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
-<a name="indentation"></a>
-### Note on Indentation
-In API Blueprint payload assets are represented by [Markdown pre-formatted code blocks][]. This means **an asset has to be indented by four spaces** relative to its existing indentation level. In the case of Markdown list items, which are also (by definition) indented, it is even one more level further making it a total of **eight spaces** or whopping twelve spaces for list items nested under ano ...
+You can find an implementation of this API at `http://polls.apiblueprint.org/` along with the complete [Poll API Blueprint][] in the [API Blueprint Examples][] repository, or you can enjoy it [rendered on Apiary][].
```
I would break off the last part into a new sentence. 

> ...repository. You can also enjoy it...

smizell commented:
@kylef Small fixes. I have a few more final thoughts after these, but getting super close.

smizell commented:
@kylef @zdne My only other comment on this is that I think it may be helpful to provide some signifier that you are gradually building a blueprint when they see the first blueprint. I'm not sure it's clear that when you get the API Blueprint heading that you are not seeing the full blueprint.

I also never get to actually see the complete blueprint in this tutorial, as I have to click into another page. This may be by design, but it requires that I fit all of these pieces together in my head as I read the tutorial. It almost seems like you want to show the full thing then say we are now going to break this down. Just a thought.

Other than that, looks good :) I'll skim it one last time.

smizell commented on Tutorial.md:
```diff
@@ -11,477 +10,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+### Actions
 
-> **NOTE:** Resource name is an [API Blueprint Identifier][] and as such it can be composed only from a combination of alphanumerical characters, underscores, dashes and a spaces.
+API Blueprint allows you to specify the actions you can make on each resource. As a sub-heading inside of the resource with a title, followed by the HTTP method for the action.
```
This last sentence is still not a full sentence.

smizell commented on Tutorial.md:
```diff
@@ -11,477 +10,247 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
+### Create a New Question [POST]
 
-    The Gist resource has the following attributes:
+You may create your own question using this action. It takes a JSON object containing a question and a collection of answers in the form of choices.
```
Will this tutorial be hosted on Github or will it be generated somewhere? Wondering if you should add a line break in this line as it runs off the page (since it's highlighted as code).

kylef replied:
It will be hosted on http://apiary.io/blueprint. It might make sense to line break quoted markdown, but the problem lies that it wont be perfect as it really depends on the readers resolution. If they're on a small resolution etc.

smizell commented:
@kylef Ok, had two more little typos and a comment. Looks good!

kylef commented:
Thanks @smizell, I understand where you are coming but I don't think adding the full blueprint to the top of the tutorial will help. It will result in many page scrolls before a reader will get into the tutorial ([preview](https://gist.github.com/kylef/bf832ba5c90aecb3e72f)). Would perhaps mentioning the full version with a hyperlink at the top of the tutorial solve this problem for you?

smizell commented:
@kylef That could be a help I think. Something to let them know the first thing they see isn't the full blueprint but rather a part of one.

zdne commented on Tutorial.md:
```diff
@@ -11,477 +10,249 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
-    - created_at
-    - description
-    - content
+- question (string) - The question
```
@kylef I am still uneasy with this. I understand your point on mixing `-`s and `+`s and I am fine with it. But I think we should have this unified in the tutorial to not confuse newcomers. 

Also, since we do not expect any upfront Markdown knowledge I would expect explanation of lists somewhere (-/+/*). Markdown list is a building stone of API Blueprint syntax..

zdne commented:
I like it. I think we are getting to a very direct, minimal yet complete tutorial. @kylef please address @smizell remarks and my on lists.

kylef commented:
@smizell Can you take a read of the first paragraph again and see if it addresses your concerns.

smizell commented:
@kylef I think that's a big help. This is just a thought for improvement, but I'm wondering if you could add something to key the user of the first step. You do say "The start of..." but you could expand that slightly to say something like, "The first step for creating the blueprint is to define an API Name and metadata. This step looks as follows." 

I'm not sure that's necessary, but the first paragraph and the "API Blueprint" section feel a tad disconnected. Maybe there is a simple way to connect them?

Other than that small idea, I think this looks great. Nice work!

kylef commented:
Thanks, added something very close to that :+1:.

smizell commented:
:+1: Looks great. That helps.

zdne commented on Tutorial.md:
```diff
@@ -1,7 +1,6 @@
-Welcome to the API Blueprint Tutorial! In this tutorial you will learn all about all the key features of the API Blueprint Language on the example of an API for our imaginary industry-changing Gist service.
 
-Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set some expectations let's provide following functionality in our API: *List Gists, Retrieve a Single Gist, Create a Gist, Edit a Gist, Star a Gist, Unstar a Gist, Check if a Gist is starred and Delete a Gist.*
+Welcome to an API Blueprint Tutorial! This tutorial will take you though the basics of the API Blueprint language. We’re going to build an API Blueprint step by step for a service called Polls – a simple API allowing consumers to view polls and vote in them. You can take a look at the [full version][Poll API Blueprint] of the blueprint used in this tutorial for reference.
```
Minor

> We’re going to build an API Blueprint ...

I would say the Blueprint should be with lowercase `b` since we are building a blueprint file not Blueprint language.

zdne commented on Tutorial.md:
````diff
@@ -11,477 +10,251 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
-    + Response 200
-
-        [Star][]
+```
````
Missing Markdown syntax highlighting

zdne commented on Tutorial.md:
````diff
@@ -11,477 +10,251 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
-        [Star][]
+```
+### Delete [DELETE]
++ Response 204
````
Minor: 

For consistency I would say you want a newline between header and the list item

zdne commented on Tutorial.md:
```diff
@@ -11,477 +10,251 @@ Gist is a paste. A snippet of text. Not unrelated to [GitHub Gists][].  To set s
 
-    # Group Gist
-    Gist-related resources of *Gist Fox API*.
+> **Note:** Lists are created by preceding list items with either a `+`, `*` or `-`.
```
I would expect a bigger "list" introduction (one sentence will do). Possibly on its first use – when you are explaining action's response?

zdne commented:
@kylef updated review. I think we are almost there, hold on! :dancer:

kylef commented:
:balloon: :tada: :balloon: :tada: :balloon: :tada: :balloon: :tada: :balloon: :tada: 

![](http://media.giphy.com/media/cIQkxw4G65TG/giphy.gif)

zdne commented:
Ha ha
